### DISCUSSION TITLE
Option to make the Contribution Scores more Interpretable

### DISCUSSION THREAD

**Discussion by joshdunnlime** [interpretml/gam-changer#3] [General]
In several domains, Contribution Scores/Shape Functions have real world meaning. It would be useful to visualise these learned shape functions on their correct scale.

For example, when using an EBM to predict the effects of a individual Air Handler Unit's (AHU) heating valve on the energy used by an attached boiler system, these shape fucntions approximate the efficiency of the AHU. E.g. for each 1% increase in valve openness, 0.2kWh increase in energy usage. However, as you can see, the predicted values are in kWh (energy). In this domain, negative energy doesn't make sense, and then using 0% valve openess, we would expect 0 extra heat usage. Therefore, the ability to rescale the shape function so that it is non-negative (in this case) would be really useful for stateholder understanding and model interpretation.

**Comment by joshdunnlime** [comment]
I will add plots with examples when I get the chance. Currently, I don't have my example to hand.

**Comment by xiaohk** [comment]
Agreed!

We conducted a user study for GAM Changer last year: https://github.com/interpretml/interpret/issues/283 (we would have learned a lot from you if we had recruited you! 😁). One participant told us something similar: the predicted score should not be negative due to the definition of the value in their task. Therefore, they found the alignment and delete tool helpful, as they can easily "rescale" the scores in the negative region to be 0.

With GAM Changer you can rescale the shape function to be non-negative. Does GAM Changer meet your need for this example? Or you think there can be more sophisticated methods to recalibrate contribution scores?

Interpreting negative scores should be careful though. If you use multiple features to predict AHU, the negative score region on one feature should be interpreted in the context of considering all other features. For example, the score on `valve openness` can be negative if data in this region has larger positive scores on other features. With GAM Changer, you can also check correlated features with the feature panel. I would suggest select the negative region on `valve openness` and see if there are some "suspicious" features popping up in the feature panel.

https://user-images.githubusercontent.com/15007159/153772516-3043ae5d-5709-4381-a14d-be4c3cb89253.mp4

**Comment by xiaohk** [comment]
Thanks for raising this issue. I think it would be a good topic to discuss, and I will convert this issue into a discussion.

**Comment by joshdunnlime** [comment]
Using your example above,we might wish to do something like this: ![Screenshot 2022-02-14 at 11 23 22 1](https://user-images.githubusercontent.com/89857735/153855622-5af5672b-122f-4f2f-bc9d-67230c7f3ced.png)

![Screenshot 2022-02-14 at 11 25 46](https://user-images.githubusercontent.com/89857735/153861114-722358d9-6537-4f6a-a263-2fb4912f5ce7.png)

I have a question however - If I was to make this kind of change to the model, would this increase in model score (for this vairiable) be remove from the intercept term? When I was using this on my AHU example, if I shifted a term up by 40, I would need to subtract 40 from the intercept to make sure the model didn't introduce any bias or reduce model accuracy, for example.

**Reply by xiaohk** [reply]
Oh I see. After training an EBM model, each graph is intentionally centered around 0. It might explain the strange negative region in your shape functions. Internally, the `interpret` library vertically shifts each graph and adjusts the intercept constant accordingly so that the mean prediction for each feature is 0 across the training data and the model "remains intact". You can learn more from this issue: https://github.com/interpretml/interpret/issues/207

It makes sense that in your case you want to reverse this process. GAM Changer does not adjust the intercept constant. Therefore, you would need to manually adjusts the scores of your features as well as the intercept constant. For example, you can add 0.5 to all scores of `GrLivArea`, and you would also need to subtract 0.5 from the intercept term.

**Reply by xiaohk** [reply]
The "centering" code is here:

https://github.com/interpretml/interpret/blob/755963a767f2761a87f7f00345fc1532b8c66ffd/python/interpret-core/interpret/glassbox/ebm/utils.py#L221-L227

I don't think there is a way to disable this procedure when you train an EBM. You can raise an issue in [interpret/interpret](https://github.com/interpretml/interpret/issues) if you would like to disable it.
